lib/src/source_visitor.dart

```diff
@@ -3205,6 +3160,65 @@ class SourceVisitor extends ThrowingAstVisitor {
     if (nest) builder.unnest();
   }
 
+  /// Visits an infix operator-like AST node: a binary operator expression, or
+  /// binary pattern.
+  ///
+  /// Applies the same formatting to all. In particular, it flattens out a
+  /// series of the same operator precedence so that if one splits, they all do.
+  ///
+  /// [T] is the type of node being visited and [destructureNode] is a callback
+  /// that takes one of those and yields the operands and operator. We need
+  /// this since there's no interface shared by the various binary operator
+  /// AST nodes.
```

Comment:
I'm trying to understand the implications of the last sentence. Can a `LogicalAndPattern` be an operand in a `LogicalOrPattern` or vice versa?

Why do we _only_ care to treat `LogicalAndPattern` specially  when it is recursively under another `LogicalAndPattern` as the entrypoint?

Can you expand this doc comment to so that it is clear why it is useful to recurse through `traverse` specifically for the T -> T nesting case, while non-T go through `visit`?

Reply:
Am I correct that both `LogicalAndPattern` and `LogicalOrPattern` were previously modeled as the single type `BinaryPattern`? If those _can_ nest I don't think this will behave the same unless each call knows how to handle both of those types.

Reply:
> Am I correct that both `LogicalAndPattern` and `LogicalOrPattern` were previously modeled as the single type `BinaryPattern`?

That's correct. But they have different precedences so they behave the same with this change as they did before. A LogicalOrPattern should only traverse and flatten out the other LogicalOrPatterns in its operands. Any LogicAndPatterns (which are higher precedence) should do their own internal flattening. For example:

```dart
if (object
    case first && second ||
        third &&
            fourth &&
            fifth &&
            sixth ||
        seventh) {
  ;
}
```

All of the `||` form a single flattened chain and we split at all of them even though `seventh` could fit on the previous line. Then, within those, each series of `&&` forms its own independent chain. The `first && second` one doesn't split, but the `third ... sixth` one does. I'll see if I can word the doc better.

Reply:
> Now that they're different AST types, that check isn't needed. (It is still needed for `BinaryExpression`, though.)

Ah ok, that does clear up my confusion. I noticed that the precedence check wasn't applied, it makes sense that it's because we know that the precedence and type are 1:1 for that case.